## Purpose

`cmuts normalize` is designed to take modification counts generated by `cmuts core`, and generate normalized reactivity profiles for each RNA in the probing library. It will also generate additional coverage statistics and figures for visualization of the generated profiles.

In the case that `cmuts core` was run with the `--pairwise` argument, the 2D counts computed will also be processed, in order to compute modification rate correlations and similar pairwise statistics.

## Usage

To run `cmuts normalize` requires an HDF5 file of modification counts and the original sequence library in FASTA format. The basic syntax is
```bash
cmuts normalize \
  -o "$OUTPUT" \
  --mod "$MOD" \
  --nomod "$NOMOD" \
  --fasta "$FASTA"
  $FILE
```
The `--mod` and `--nomod` flags specify the HDF5 datasets in which the treated and untreated modification counts are stored, respectively. The latter is optional.

!!! tip
    More than one file can be passed to `cmuts normalize`. In such a case, the modification counts are summed across the files before normalization. This is useful in the case where the inputs to `cmuts core` were split across multiple files.

## Command Line Options

### Core Options

**`-o, --output`** : Output HDF5 filename (required)

**`--mod`** : Name of the dataset containing treated modification counts (required)

**`--nomod`** : Name of the dataset containing untreated modification counts

**`--fasta`** : FASTA file of probed RNA sequences.


### Output Control

**`--overwrite`** : Overwrite existing HDF5 file

**`--group`** : HDF5 group to place the output datasets in (none if not specified)


### Profile Computation

**`--no-insertions`** : Do not use insertions to compute the reactivity profiles

**`--no-deletions`** : Do not use deletions to compute the reactivity profiles


### Normalization Control

**`--raw`** : Do not normalize the reactivity values

**`--clip-low`** : Clip negative reactivity values

**`--clip-high`** : Clip reactivity values above 1

**`--blank-5p`** : NaN out this many bases on the 5' end (default: 0)

**`--blank-3p`** : NaN out this many bases on the 3' end (default: 0)

**`--blank-cutoff`** : NaN out any positions with less than this many reads (default: 10)

**`--norm-independent`** : Normalize each profile separately, rather than using experiment-wide statistics 


### Pairwise Settings

**`--sig`** : The significance value to use when plotting pairwise modificaiton correlations.


## Outputs

### Files

The output of `cmuts normalize` will be an HDF5 file with the following structure:

```
/
├── roi-mask
├── SNR
├── error
├── heatmap
├── norm
├── reactivity
└── reads
```

**roi-mask** (Region Of Interest): A boolean array indicating the region of the RNA which is of structural importance (i.e. all but the flanking sequences and low-coverage regions).

**SNR** (Signal-To-Noise): An estimate of the average SNR per nucleotide, based on the computed error, one for each reference.

**error**: The standard error of the mean at each nucleotide.

**heatmap**: The prevalence of each mutation, insertion, deletion, and termination type throughout the entire library.

**norm**: The normalization value used. It will be a scalar unless `--norm-independent` is passed, in which case each reference has its own norm value.

**reactivity**: The reactivity profiles.

**reads**: The number of reads used to compute each reactivity profile.

In the case where 2D data was also present in the input, the output HDF5 file will also contain

```
/
├── correlation
└── mutual-information
```

!!! warning
    If a value was passed to `--group`, then the above will be contained in an HDF5 group with that name.

### Figures
